pkg/volume/awsebs/aws_util.go

```diff
@@ -287,3 +288,53 @@ func findNvmeVolume(findName string) (device string, err error) {
 
 	return resolved, nil
 }
+
+func formatVolumeID(volumeID string) (string, error) {
+	// This is a workaround to fix the issue in converting aws volume id from globalPDPath and globalMapPath
+	// There are three formats for AWSEBSVolumeSource.VolumeID and they are stored on disk in paths like so:
+	// VolumeID									mountPath								mapPath
+	// aws:///vol-1234					aws/vol-1234						aws:/vol-1234
+	// aws://us-east-1/vol-1234 aws/us-east-1/vol-1234  aws:/us-east-1/vol-1234
+	// vol-1234									vol-1234								vol-1234
+	// This code is for converting volume ids from paths back to AWS style VolumeIDs
+	sourceName := volumeID
+	if strings.HasPrefix(volumeID, "aws/") || strings.HasPrefix(volumeID, "aws:/") {
+		names := strings.Split(volumeID, "/")
+		length := len(names)
+		if length < 2 || length > 3 {
+			return "", fmt.Errorf("invalid volume name format %q", volumeID)
+		}
+		volName := names[length-1]
+		if !strings.HasPrefix(volName, "vol-") {
+			return "", fmt.Errorf("Invalid volume name format for AWS volume (%q)", volName)
+		}
+		if length == 2 {
+			sourceName = awsURLNamePrefix + "" + "/" + volName // empty zone label
+		}
+		if length == 3 {
+			sourceName = awsURLNamePrefix + names[1] + "/" + volName // names[1] is the zone label
+		}
+		klog.V(4).Infof("Convert aws volume name from %q to %q ", volumeID, sourceName)
+	}
+	return sourceName, nil
+}
+
+func newAWSVolumeSpec(volumeName, volumeID string, isBlock bool) *volume.Spec {
+	block := v1.PersistentVolumeBlock
+	awsVolume := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				AWSElasticBlockStore: &v1.AWSElasticBlockStoreVolumeSource{
+					VolumeID: volumeID,
+				},
+			},
+		},
+	}
+	if isBlock {
+		awsVolume.Spec.VolumeMode = &block
```

Comment:
Aparently the mode must be set also for FS volumes, otherwise I get this error during FS volume reconstruction:

`
W1009 12:00:50.644436    6566 reconciler.go:394] Could not construct volume information, cleanup the mounts. (pod.UID 199d1aa9-3f21-438b-8286-d84b0cc0e774, volume.SpecName pvc-44f6877f-63e9-41aa-b301-406c83778dfb): cannot get volumeMode for volume: pvc-44f6877f-63e9-41aa-b301-406c83778dfb
`

Reply:
done